src/rust/catnap/queue.rs

```diff
@@ -81,131 +83,248 @@ impl CatnapQueue {
-    }
-
-    /// Begins `listen()` operation.
-    pub fn prepare_listen(&self) -> Result<(), Fail> {
-        self.socket.borrow_mut().prepare_listen()
+    pub fn bind(&self, local: SocketAddrV4) -> Result<(), Fail> {
+        let mut socket: RefMut<Socket> = self.socket.borrow_mut();
+        socket.prepare_bind()?;
+        match socket.bind(local) {
+            Ok(_) => {
+                socket.commit();
+                Ok(())
+            },
+            Err(e) => {
+                socket.abort();
+                Err(e)
+            },
+        }
     }
 
     /// Sets the target queue to listen for incoming connections.
-    pub fn listen(&mut self, backlog: usize) -> Result<(), Fail> {
-        self.socket.borrow_mut().listen(backlog)
+    pub fn listen(&self, backlog: usize) -> Result<(), Fail> {
+        let mut socket: RefMut<Socket> = self.socket.borrow_mut();
+        socket.prepare_listen()?;
+        match socket.listen(backlog) {
+            Ok(_) => {
+                socket.commit();
+                Ok(())
+            },
+            Err(e) => {
+                socket.abort();
+                Err(e)
+            },
+        }
     }
 
-    /// Prepare to begin accepting connections.
-    pub fn prepare_accept(&self) -> Result<(), Fail> {
-        self.socket.borrow_mut().prepare_accept()
+    /// Starts a coroutine to begin accepting on this queue. This function contains all of the single-queue,
+    /// synchronous functionality necessary to start an accept.
+    pub fn accept<F: FnOnce(Yielder) -> Result<TaskHandle, Fail>>(&self, insert_coroutine: F) -> Result<QToken, Fail> {
+        let mut socket: RefMut<Socket> = self.socket.borrow_mut();
+        // Check whether accepting on this queue is valid.
+        socket.prepare_accept()?;
+
+        // Create a yielder for the coroutine and insert it into the scheduler.
+        let yielder: Yielder = Yielder::new();
+        let yielder_handle: YielderHandle = yielder.get_handle();
+        let task_handle: TaskHandle = match insert_coroutine(yielder) {
+            Ok(handle) => {
+                // Commit to connecting this queue.
+                socket.commit();
+                handle
+            },
+            Err(e) => {
+                // Roll back connecting this queue.
+                socket.abort();
+                return Err(e);
+            },
+        };
+        self.add_pending_op(&task_handle, &yielder_handle);
+        Ok(task_handle.get_task_id().into())
     }
 
-    /// Begins the accepted operation.
-    pub fn prepare_accepted(&self) -> Result<(), Fail> {
-        self.socket.borrow_mut().prepare_accepted()
-    }
+    /// Asynchronously accepts a new connection on the queue. This function contains all of the single-queue,
+    /// asynchronous code necessary to run an accept and any single-queue functionality after the accept completes.
+    pub async fn do_accept(&self, yielder: Yielder) -> Result<Self, Fail> {
```

Comment:
Curious, why was this renamed from `accept()` to `do_accept()`? Was it because this is an `async` function? If that's the case, would it be better to call it `async_accept()` instead?

- this comment is applicable to ther do_* cases as well

Reply:
Yeah, the issue is that we currently have a close and an async_close. I could also call it accept_coroutine? I was actually using an async convention that we used in TAPIR, where the actual code to run an async operation is in a function called do_*

Reply:
Ok, got it. If we follow the same convention throughout, then do_* sounds good to me.